sw/n64/testrom/shell.c

```diff
@@ -0,0 +1,188 @@
+
+/* Colors */
+color_t MENU_BAR_COLOR = { .r = 0x82, .g = 0x00, .b = 0x2E, .a = 0x00 }; // 0x82002E, Berry
+color_t SELECTION_COLOR = { .r = 0x00, .g = 0x67, .b = 0xC7, .a = 0x00 }; // 0x0067C7, Bright Blue
+
+/* Assume default font size*/
+int calculate_num_rows_per_page() {
+    // ---top of screen---
+    // menu bar
+    // padding + more files above indicator space
+    // list of files...
+    // ...
+    // more file indicator space
+    // ---bottom of screen---
+
+    // Start by subtracting the menu bar height, and paddings from the total height
+    int availableHeight = SCREEN_HEIGHT - MENU_BAR_HEIGHT - (ROW_HEIGHT * 2);
+
+    // Since there is currently no other dynamic portion, we can use a simple calculation to find rows that will fit in the space
+    int rows = availableHeight / ROW_HEIGHT;
+
+    return rows;
+}
+
+/*
+ * Render a list of strings and show a caret on the currently selected row
+ */
+void render_list(display_context_t display, char *list[], int currently_selected, int first_visible, int max_on_screen) {
+
+    /* If we aren't starting at the first row, draw an indicator to show more files above*/
+    if (first_visible > 0) {
+        graphics_draw_text(display, MARGIN_PADDING, MENU_BAR_HEIGHT, "...");
+    }
+
+    for (int i = 0; i < max_on_screen; i++) {
+        int row = first_visible+i;
+        int x = currently_selected == row ? MARGIN_PADDING : 0; // if the current item is selected move it over so we can show a selection caret
+        int y = i * ROW_HEIGHT + LIST_TOP_PADDING; // First row must start below the menu bar, so add that plus a pad
+        
+        // if we run out of items to draw on the screen, don't draw anything else
+        if (row > NUM_ENTRIES) {
+            break;
+        }
+
+        if (currently_selected == row) {
+            /* Render selection box */
+            graphics_draw_box(display, 0, y-2, ROW_SELECTION_WIDTH, 10, graphics_convert_color(SELECTION_COLOR));
+
+            /* Now render selection caret */
+            graphics_draw_text(display, MARGIN_PADDING/2, y, "  >");
+
+            /* Increase Margin Padding to account for added caret width */
+            x += MARGIN_PADDING;
+        } else {
+            x += MARGIN_PADDING;
+        }
+
+        // Render the list item
+        graphics_draw_text(display, x, y, list[row]);
+
+        /* If this is the last row and there are more files below, draw an indicator*/
+        if (row+1 < NUM_ENTRIES && i + 1 >= max_on_screen) {
+            graphics_draw_text(display, MARGIN_PADDING, y + 10, "...");
+        }
+    }
+}
+
+void draw_header_bar(display_context_t display, int fileCount) {
+    int x = 0, y = 0, width = SCREEN_WIDTH, height = MENU_BAR_HEIGHT;
+    graphics_draw_box(display, x, y, width, height, graphics_convert_color(MENU_BAR_COLOR));
+    
+    char menuHeaderBuffer[50];
+    sprintf(menuHeaderBuffer, "PicoCart64 OS b2022.07.17\t\t\t\t%d Files\0", fileCount);
```

Comment:
Is the \0 indentional, if so, what is it for?

Reply:
It's a null terminating character; because the buffer is larger than the text I thought I might need it to avoid printing garbage but it's not actually needed. It's been a while since I have done large scale C programming and I'm paranoid that I'll bork a memory address or something lol.

Reply:
`sprintf` writes the null character in the end, so it isn't needed. I'd prefer if we used `snprintf(buf, sizeof(buf), fmt, ...)` though.

Reply:
You know I removed it but when the indent tool deleted the code, I just discarded changes via git and forgot to remove the terminating character again 🤦‍♀️ 

I'll remove that and change the method to use snprintf